driver/src/main/java/org/neo4j/driver/internal/net/pooling/BlockingPooledConnectionQueue.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.internal.net.pooling;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import org.neo4j.driver.internal.util.Supplier;
+
+/**
+ * A blocking queue that also keeps track of connections that are acquired in order
+ * to facilitate termination of all connections.
+ */
+public class BlockingPooledConnectionQueue
+{
+    /** The backing queue, keeps track of connections currently in queue */
+    private final BlockingQueue<PooledConnection> queue;
+
+    /** Keeps track of acquired connections */
+    private final Set<PooledConnection> acquiredConnections = ConcurrentHashMap.newKeySet();
+
+    public BlockingPooledConnectionQueue( int capacity )
+    {
+        this.queue = new LinkedBlockingQueue<>( capacity );
+    }
+
+    /**
+     * Offer a connections back to the queue
+     *
+     * @param pooledConnection the connection to put back to the queue
+     * @return <code>true</code> if connections was accepted otherwise <code>false</code>
+     */
+    public boolean offer( PooledConnection pooledConnection )
+    {
+        boolean offer = queue.offer( pooledConnection );
```

Comment:
Ref: https://github.com/neo4j/neo4j-dotnet-driver/blob/1.1/Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs
We used lock because we do not have a concurrent container before. We probably could have a lock free impl for java driver